src/wikitextprocessor/core.py

```diff
@@ -1909,3 +1918,14 @@ def create_strip_marker(self, node: str, content: str) -> str:
                 self.strip_marker_cache["preprocess"] += 1
                 self.strip_marker_cache[content] = num
         return f"""\x7f'"`UNIQ--{node}-{num_str}-QINU`"'\x7f"""
+
+
+def detect_expand_template_loop(stack: list[str]) -> bool:
+    stack_len = len(stack)
```

Comment:
Do you think we could add a couple of guard clauses here? Like `if stack[-1] not in stack[:-1]: return False`? We could also return True when stack_len is small, and only start checking when it grows too big.

Reply:
Short list also could have loop and check it is fast. I'll add the "if" check.

Reply:
Maybe there is a better algorithm for this function, I'm also not sure how this would impact performance.

Reply:
We can merge now that it's working, see how much performance is affected and try to figure out something later.